### PR TITLE
Bug 1800: only open terminfo once

### PR DESCRIPTION
**Pull request by rbtcollins**
I did a slightly more than minimal fix here, promoting the
term::Terminal trait to be the contract that our markdown formatting
is written to, making the auto-disabling of formatting on non-tty's
and handling of non-colour terminals be clearly separated out.

The handling of corrupt or missing terminfo configuration is also
pushed down to the underlying terminal trait object rather than
having a bypass in our higher level code; this makes the upper layers
easier to reason about.

There is a modest increase in size, as the full term::Terminal trait
is now implemented for the translation layer; a decorator macro could
probably be writtent to shrink that if needed.

There is a remaining direct use of term in download_tracker that causes
a second open of the terminfo database, which should also be fixed but
it is constant in nature so a later pass can do it.

I don't have any tests for this; doing a unit test would be
uninteresting in this case - rust's types give use what we need at that
level, and for the conceptual level 'nothing else in the system calls
any term calls', unit tests are inappropriate. A good way to test this
would be to have the functional tests instrumented and make assertions
about the behaviour of those executions; then any future regressions
(e.g. if switching to a different term library) would be detected.

### CODE REVIEW

**Comment by rbtcollins** [comment]
I'm not sure why the tests are failing; they are passing for me :/.

**Comment by kinnison** [comment]
All the failures appear to be in `cli-exact` which is ultra-sensitive to changes in output format.  it's possible there are newline or whitespace changes which are affected by your change, but I'll have to look harder to find out.  I've kicked off a test run on my laptop to see if I can spot any difference.

**Comment by kinnison** [comment]
I can confirm i see the same errors on my laptop so I'll try to diagnose later today.

**Comment by kinnison** [comment]
Looks like there're formatting codes present in the output even under test...

I changed print_indented in clitools.rs to debug-print things and:

```
out.stdout (3 lines):
    "\n      nightly-x86_64-unknown-linux-gnu unchanged\u{1b}(B\u{1b}[m - 1.3.0 (hash-n-2)\n    \n    "
out.stderr (1 lines):
    "info: \u{1b}(B\u{1b}[msyncing channel updates for \'nightly-x86_64-unknown-linux-gnu\'\n    "
expected.stdout (3 lines):
    "\n      nightly-x86_64-unknown-linux-gnu unchanged - 1.3.0 (hash-n-2)\n    \n    "
expected.stderr (1 lines):
    "info: syncing channel updates for \'nightly-x86_64-unknown-linux-gnu\'\n    "
```

**Comment by rbtcollins** [comment]
Thanks; that will be a bug in the isatty detection glue then I guess - as the tests are passing for me under vscode on windows. Will look in ~48 hours.

File: src/cli/term2.rs
Comment (kinnison) [inline src/cli/term2.rs:83]: Given this is unused, is it worth keeping in?

File: src/cli/term2.rs
Comment (kinnison) [inline src/cli/term2.rs:90]: Ditto query about unused?

File: src/cli/term2.rs
Comment (kinnison) [inline src/cli/term2.rs]: s/termporarily/temporarily/ -- seems sad to leave this typo when you corrected the one about isatty :D

File: src/cli/term2.rs
Comment (kinnison) [inline src/cli/term2.rs:209]: This *should* in reality only end up with `\r` at a guess.

File: src/cli/main.rs
Comment (kinnison) [inline src/cli/main.rs:25]: Just for cleanliness of commits, would it make sense to split this module out as one commit, and then do the caching functionality as a second commit?  If you think it's too much of a pain then I won't block on this, but I think I'd prefer if the entirety of `markdown.rs` was not blamed on caching terminfo :D

File: src/cli/rustup_mode.rs
Comment (kinnison) [inline src/cli/rustup_mode.rs:811]: Should this not be `term2::StdoutTerminal` to reduce leakage of `term` api?

**Review by kinnison — CHANGES_REQUESTED** [review]
Mostly LGTM.  A few comments / tweaks, and then I think we could merge.  I like the caching, I like the split out of the markdown stuff (particularly since we've had some discussion about moving to a different markdown crate in the future too)

File: src/cli/term2.rs
Comment (rbtcollins) [reply src/cli/term2.rs:83]: Ehh. If we were keeping term for sure; and actually as its a closer fit to the termcolor api I think it shows the sort of change that may be needed again here, so yes.

File: src/cli/term2.rs
Comment (rbtcollins) [reply src/cli/term2.rs]: ta

File: src/cli/term2.rs
Comment (rbtcollins) [reply src/cli/term2.rs:209]: I think its ok; I think the comment is probably more helpful than someone who *does* find a problem going 'wtf' ;)

File: src/cli/main.rs
Comment (rbtcollins) [reply src/cli/main.rs:25]: Its a little awkward because the types were comingled; so we can't not change the md stuff. TBH I think having it in the one commit isn't going to make bisection a problem - and thats all I'd really worry about history wise.

File: src/cli/rustup_mode.rs
Comment (rbtcollins) [reply src/cli/rustup_mode.rs:811]: I don't think so; thats a concrete type with the isatty and error silencing built in (which BTW is really showing where we abuse the API and call e.g. reset() without checking supports_reset() .... but thats a different story); anyhow - point is that print_header doesn't care about that distinction, and it was giving it the concrete type in the past that was encouraging the markdown comingling in the first place. So I want to make it clearer:

this depends on 'a terminal'

this implements our terminal specialisations.

And I think depending on term2:StdoutTerminal gets that wrong because of the current type alias.

I did have a patch where I ditched that entirely and term2:StdoutTerminal was an export of term:StdoutTerminal, and stdio() was returning a Box, but I decided that that didn't add much.

File: src/cli/term2.rs
Comment (kinnison) [reply src/cli/term2.rs:83]: Accepted

File: src/cli/term2.rs
Comment (kinnison) [reply src/cli/term2.rs:209]: Accepted.

File: src/cli/main.rs
Comment (kinnison) [reply src/cli/main.rs:25]: Okay, yeah, looking through again I agree it'd be a pain.

File: src/cli/rustup_mode.rs
Comment (kinnison) [reply src/cli/rustup_mode.rs:811]: Aha, right, could we have an `impl` in there then (i.e. `&mut impl term::StdoutTerminal`) just to make it clear that this is a trait not a struct we're getting a borrow to?

Comment (rbtcollins) [reply src/cli/rustup_mode.rs:811]: mm, uh,so actually it is a struct we're passing in, not a trait object, so I dunno. Can we just declare it bankrupt as we're about to rip out term anyway?

Comment (kinnison) [reply src/cli/rustup_mode.rs:811]: Oh, argh, yeah OK

**Review by kinnison — APPROVED** [review]

**Comment by rbtcollins** [comment]
What needs to happen to merge this?

**Comment by jens1o** [comment]
I'd be also interested in this, so I can really start working on #1818 without having merge conflicts in the end :/

**Comment by rbtcollins** [comment]
I have a followup that will likely conflict with 1818 as well; - the reason download_tracker didn't use term2 was that term2 was in the cli package, but the download_tracker was in the root, and I presume we don't want the root depending on the CLI.

**Comment by rbtcollins** [comment]
(that branch is https://github.com/rbtcollins/rustup.rs/pull/new/tracker-term-use ) - I'll set up the PR once this merges.

**Comment by kinnison** [comment]
Sorry for the delay :D

**Pull request merged**
